### ISSUE TITLE
deprecated attribute detection clang/icc

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#6435]
I have a setup where the test DEAL_II_COMPILER_HAS_CXX14_ATTRIBUTE_DEPRECATED passes (code compiles without error), but I am getting the following warnings during compilation:
```
"warning: use of the 'deprecated' attribute is a C++14 extension" (clang6 in c++11 mode)
"warning #1292: unknown attribute "deprecated"" (icc 2018)
```
This triggers hundreds of times of course.

I tried fixing this in cmake/checks/check_01_compiler_features.cmake by adding ``PUSH_CMAKE_REQUIRED("-Werror")``, but then the test always fails with the deprecation warning "old_fn is deprecated" in the line ``int (*fn_ptr)() = old_fn;`` because this is also treated as an error. If I now remove the ``int (*fn_ptr)() = old_fn;`` line (to remove the deprecation warning), the warnings above no longer trigger. I would need to add ``-Wall -pedantic -Wextra --std=c++11 -Wno-deprecated-declarations -Wno-deprecated -Werror`` for the test to work correctly.

Thoughts?

**Comment by tjhei** [comment]
> The right way to do it is to switch the order of the checks

This works for clang but it is still not a robust check: we check if we support the flag with different compiler flags than the ones that are later used to compile the code. This includes ``pedantic`` and the c++ standard.

**Comment by drwells** [comment]
For the record: I also tried the `-Werror` guard first/

**Comment by tjhei** [comment]
my try at fixing things: #6438

**Comment by tjhei** [comment]
closed by #6438
